pkg/pss/pss.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package pss
+
+import (
+	"context"
+	"sync"
+
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/pushsync"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/tags"
+	"github.com/ethersphere/bee/pkg/trojan"
+)
+
+// Pss is the top-level struct, which takes care of message sending
+type Pss struct {
+	pusher     pushsync.PushSyncer
+	tags       *tags.Tags
+	handlers   map[trojan.Topic]Handler
+	handlersMu sync.RWMutex
+	metrics    metrics
+	logger     logging.Logger
+}
+
+type Options struct {
+	Logger     logging.Logger
+	PushSyncer pushsync.PushSyncer
+	Tags       *tags.Tags
+}
+
+// NewPss inits the Pss struct with the storer
+func NewPss(o Options) *Pss {
+	return &Pss{
+		pusher:   o.PushSyncer,
+		tags:     o.Tags,
+		handlers: make(map[trojan.Topic]Handler),
+		metrics:  newMetrics(),
+		logger:   o.Logger,
+	}
+}
+
+// Handler defines code to be executed upon reception of a trojan message
+type Handler func(trojan.Message)
+
+// Send constructs a padded message with topic and payload,
+// wraps it in a trojan chunk such that one of the targets is a prefix of the chunk address
+// uses push-sync to deliver message
+func (p *Pss) Send(ctx context.Context, targets trojan.Targets, topic trojan.Topic, payload []byte) (*tags.Tag, error) {
+	p.metrics.TotalMessagesSentCounter.Inc()
+
+	//construct Trojan Chunk
+	m, err := trojan.NewMessage(topic, payload)
+	if err != nil {
+		return nil, err
+	}
+	var tc swarm.Chunk
+	tc, err = m.Wrap(targets)
+	if err != nil {
+		return nil, err
+	}
+
+	tag, err := p.tags.Create("pss-chunks-tag", 1, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// push the chunk using push sync so that it reaches it destination in network
+	if _, err = p.pusher.PushChunkToClosest(ctx, tc.WithTagID(tag.Uid)); err != nil {
+		return nil, err
+	}
+
+	tag.Total = 1
+
+	return tag, nil
+}
+
+// Register allows the definition of a Handler func for a specific topic on the pss struct
+func (p *Pss) Register(topic trojan.Topic, hndlr Handler) {
+	p.handlersMu.Lock()
+	defer p.handlersMu.Unlock()
+	p.handlers[topic] = hndlr
+}
+
+// Deliver allows unwrapping a chunk as a trojan message and calling its handler func based on its topic
+func (p *Pss) Deliver(c swarm.Chunk) {
```

Comment:
I would suggest to add context to Deliver and Handler, and probably also to return error. Something noticed from the chunk repair pr.